enhancements/agent-installer/interactive-network-config.md

```diff
@@ -0,0 +1,238 @@
+replaces: N/A
+superseded-by: N/A
+---
+
+# Interactive Network Configuration
+
+## Summary
+
+Enable the user, after booting the agent ISO, to interactively add or correct
+any neccessary network configuration. Since we cannot rely on network
+connectivity for this step, configuration will be via a text user interface
+(TUI) displayed on the console.
+
+## Motivation
+
+Providing network configuration up-front, in the absence of the actual hosts to
+be used, is onerous for users. Names of NIC interfaces are difficult to predict
+in advance, and often MAC addresses are required. Even the simplified NMState
+format is difficult to write by hand.
+
+Correcting any error in the configuration passed into the installer currently
+requires returning to the beginning of the process, generating a new ISO and
+booting the hosts anew with the updated image. There is no guarantee of getting
+it right, so there could be multiple iterations of this process.
+
+### User Stories
+
+* As a cluster administrator, I want to install a cluster into a network
+  environment requiring custom network configuration (e.g. static IPs, NIC
+  bonds, VLANs) without needing any information from the physical servers in
+  advance.
+* As a cluster administrator, I want to fix a cluster installation that has
+  stalled due to incorrect network configuration without having to regenerate
+  the installation image.
+
+### Goals
+
+* Allow the user to configure static IPs, NIC bonds, and VLANs.
+* Make available all necessary information about the hardware and network
+  environemnt to do this configuration.
+* Provide an indication of when sufficient connectivity is or is not available.
+* Enable the user to debug connectivity problems by narrowing down the cause
+  (DNS, reachability, &c.).
+* Work for both the existing automation workflow and a future fully-interactive
+  workflow.
+
+### Non-Goals
+
+* Interactive (re)configuration of any other aspect of the installation.
+* Add interactive network configuration to CoreOS in general.
+* Allow interactive network configuration during IPI baremetal installs.
+
+## Proposal
+
+### Workflow Description
+
+The user will generate the agent ISO using `openshift-install` as usual. The
+`networkConfig` section in the `agent-config` (or the NMStateConfig ZTP
+manifests) may _or may not_ contain network configuration.
+
+The user will then boot the ISO. Any network configuration for each current
+host contained in the ISO will be applied at startup.
+
+Prior to the login prompt, a TUI will appear on the hardware console. It will
+display the current status of connectivity to the registry, which will be
+continuously updated. It will also give the user the option to interactively
+reconfigure the network, or accept the current configuration and exit.
+
+If there is connectivity available to the registry, the TUI will autmatically
+exit within 20s if there is no user interaction.
```

Comment:
Is this the same for the "non-host0" hosts? Will there be a way to recover if the issue is that the host0 host is not reachable by other hosts because of incompatible IP config or will that require a 2nd reboot?

Reply:
I'm proposing not to check for connectivity to node 0, because it limits our options for future development - both for getting this feature into CoreOS and for how we implement the interactive workflow.
In practice I suspect connectivity to the registry from each host is usually going to be sufficient for nodes to find each other. If it isn't then the user will have to log in and fix it or reboot and try again I guess.

Reply:
OK, sounds good